### PR TITLE
refactor: 00029 refactor demo to lmax disruptor and 00055 persistence design doc

### LINKED ISSUE #29: Refactor Live Stream Mediator to decouple producer from consumer
With [this PR](https://github.com/hashgraph/hedera-block-node/pull/23), the live stream of items are broadcast from the producer to all consumers via mediator that all consumers subscribe to.  There are concerns that this solution will not scale well with a larger number of consumers.  This strategy is also at odds with the framework hooks Helidon provides.  The mediator must take greater efforts to make it work.

Given these drawbacks, we should refactor the application to use an intermediate data structure to store the inbound items from the producer.  Consumers will then receive BlockItems from the intermediate data structure and pass them back to their downstream consumer.

### PR DESCRIPTION
**Pull request by mattp-swirldslabs**
**Description**:
* Integration with LMAX Disruptor RingBuffer
* Changed toy types to align with names and relationships like
  hedera-protobufs
* Boosted unit test coverage to 100% for most packages
* Adjusted Docker container to run as the hedera user rather than root
* Changed the consumer `bidirection stream` to `server streaming`
* Added exception handling when reading and writing to storage
* Refactored getBlock to singleBlock to align with `hedera-protobufs` rpc service definition and types
* Separated ReadBlock and WriteBlock
* Added RemoveBlock to handle removal of partially written Blocks when an exception is thrown
* Adjusted producer and consumer scripts to work with new types
* Added metrics to report: live block items counted, blocks persisted, single blocks retrieved, and current subscribers
* #55 

**Related issue(s)**:

Fixes #29, #55

**Notes for reviewer**:

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

**Review by AlfredoG87 — CHANGES_REQUESTED** [review]
LGTM, only thing that I would add is a Diagram on the design doc, so is easier to understand the ring buffer and the interactions between the classes.

Leaved some nits.

File: server/src/main/java/com/hedera/block/server/BlockStreamService.java [hunk 1]
```diff
@@ -94,83 +97,113 @@ public String serviceName() {
...
+
+        LOGGER.log(System.Logger.Level.DEBUG, "Executing Unary singleBlock gRPC method");
+
+        if (serviceStatus.isRunning()) {
+            final long blockNumber = singleBlockRequest.getBlockNumber();
+            try {
+                final Optional<Block> blockOpt = blockReader.read(blockNumber);
+                if (blockOpt.isPresent()) {
+                    LOGGER.log(
+                            System.Logger.Level.INFO,
```
[the remark this answers is not in the document]
Comment (mattp-swirldslabs) [reply server/src/main/java/com/hedera/block/server/BlockStreamService.java:161]: Good point, I forgot to change this back after building the feature out.  I've been using DEBUG for most of these. Do you mind if I use DEBUG here as well?

File: server/src/test/java/com/hedera/block/server/consumer/ConsumerStreamResponseObserverTest.java [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+        when(testClock.millis()).thenReturn(TEST_TIME, TEST_TIME + TIMEOUT_THRESHOLD_MILLIS + 1);
+
+        final var consumerBlockItemObserver =
+                new ConsumerStreamResponseObserver(
+                        TIMEOUT_THRESHOLD_MILLIS,
+                        testClock,
+                        streamMediator,
+                        responseStreamObserver);
+
+        Thread.sleep(51);
```
[the remark this answers is not in the document]
Comment (mattp-swirldslabs) [reply server/src/test/java/com/hedera/block/server/consumer/ConsumerStreamResponseObserverTest.java]: Good catch! I forgot to remove that when I converted to using a InstantSource mock

File: server/src/main/java/com/hedera/block/server/mediator/LiveStreamMediatorImpl.java [hunk 3]
```diff
@@ -16,84 +16,183 @@
...
     @Override
-    public void unsubscribe(final LiveStreamObserver<BlockStreamServiceGrpcProto.Block, BlockStreamServiceGrpcProto.BlockResponse> liveStreamObserver) {
-        if (subscribers.remove(liveStreamObserver)) {
-            LOGGER.log(System.Logger.Level.DEBUG, "Successfully removed observer from subscription list");
+    public void publish(final BlockItem blockItem) throws IOException {
+
+        if (serviceStatus.isRunning()) {
+
+            // Publish the block for all subscribers to receive
+            LOGGER.log(System.Logger.Level.INFO, "Publishing BlockItem: {0}", blockItem);
```
[the remark this answers is not in the document]
Comment (mattp-swirldslabs) [reply server/src/main/java/com/hedera/block/server/mediator/LiveStreamMediatorImpl.java:116]: Good catch. I forgot to change this back to DEBUG after troubleshooting. Fixed

File: server/src/main/java/com/hedera/block/server/BlockStreamService.java [hunk 4]
```diff
@@ -94,83 +97,113 @@ public String serviceName() {
...
+                .build();
+    }
+
+    static SingleBlockResponse buildSingleBlockNotFoundResponse() {
+        return SingleBlockResponse.newBuilder()
+                .setStatus(SingleBlockResponse.SingleBlockResponseCode.READ_BLOCK_NOT_FOUND)
+                .build();
+    }
+
+    // TODO: Fix this error type once we have it - should not be success
```
[the remark this answers is not in the document]
Comment (mattp-swirldslabs) [reply server/src/main/java/com/hedera/block/server/BlockStreamService.java]: Fixed.

File: server/src/main/java/com/hedera/block/server/ServiceStatusImpl.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        this.webServer = webServer;
+    }
+
+    /**
+     * Stops the service and web server. This method is called to shut down the service and the web
+     * server in the event of an unrecoverable exception or during expected maintenance.
+     */
+    public void stopWebServer() {
+        isRunning.set(false);
+        webServer.stop();
```
[the remark this answers is not in the document]
Comment (mattp-swirldslabs) [reply server/src/main/java/com/hedera/block/server/ServiceStatusImpl.java:67]: I added some additional comments. The order matters. We don't want to accept new connections before the server shuts down. This flag also triggers running connections to respond with appropriate, spec-related protobuf responses.

File: server/src/main/java/com/hedera/block/server/BlockStreamService.java [hunk 6]
```diff
@@ -94,83 +97,113 @@ public String serviceName() {
...
+                if (blockOpt.isPresent()) {
+                    LOGGER.log(
+                            System.Logger.Level.INFO,
+                            "Successfully returning block number: {0}",
+                            blockNumber);
+                    singleBlockResponseStreamObserver.onNext(
+                            buildSingleBlockResponse(blockOpt.get()));
+                } else {
+                    LOGGER.log(
+                            System.Logger.Level.DEBUG, "Block number {0} not found", blockNumber);
```
[the remark this answers is not in the document]
Comment (AlfredoG87) [reply server/src/main/java/com/hedera/block/server/BlockStreamService.java:168]: Yes I agree that either DEBUG or TRACE are appropiate, but yes, I think consistency is better, so lets use DEBUG 👍

File: server/docs/design/block-persistence.md [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+**BlockWriter** - An interface defining methods used to write BlockItems to storage. It represents a lower-level
+component whose implementation is directly responsible for writing a BlockItem to storage as a Block.
+
+**BlockRemover** - An interface defining the methods used to remove a Block from storage. It represents a lower-level
+component whose implementation is directly responsible for removing a Block from storage.
+
+---
+
+### Design
```
[the remark this answers is not in the document]
Comment (AlfredoG87) [reply server/docs/design/block-persistence.md:45]: I forgot we already have a design document for this, nvm...

**Review by AlfredoG87 — APPROVED** [review]

**Review by AlfredoG87 — DISMISSED** [review]
LGTM

**Review by AlfredoG87 — APPROVED** [review]
LGTM

**Pull request merged**